tests/functional/test_discovery.py

```diff
@@ -0,0 +1,371 @@
+            config=config,
+            endpoint_url=endpoint
+        )
+        with http_stubber as stubber:
+            stubber.add_response(status=200, body=b'{}')
+            client.test_discovery_required(Foo="bar")
+            self.assert_discovery_skipped(
+                stubber,
+                b"test-discovery-endpoint.TestDiscoveryRequired"
+            )
+            self.assert_endpoint_used(stubber.requests[0].url, endpoint)
+
+    def test_endpoint_discovery_optional_with_custom_endpoint(self):
+        endpoint = "https://custom.domain/"
+        client, http_stubber = self.create_client(
+            service_name="test-discovery-endpoint", endpoint_url=endpoint
+        )
+        with http_stubber as stubber:
+            stubber.add_response(status=200, body=b'{}')
+            client.test_discovery_optional(Foo="bar")
+            self.assert_discovery_skipped(
+                stubber,
+                b"test-discovery-endpoint.TestDiscoveryOptional"
+            )
+            self.assert_endpoint_used(stubber.requests[0].url, endpoint)
+
+    def test_endpoint_discovery_optional_disabled_with_custom_endpoint(self):
+        endpoint = "https://custom.domain/"
+        config = Config(endpoint_discovery_enabled=False)
+        client, http_stubber = self.create_client(
+            service_name="test-discovery-endpoint",
+            config=config,
+            endpoint_url=endpoint
+        )
+        with http_stubber as stubber:
+            stubber.add_response(status=200, body=b'{}')
+            client.test_discovery_optional(Foo="bar")
+            self.assert_discovery_skipped(
+                stubber,
+                b"test-discovery-endpoint.TestDiscoveryOptional",
+            )
+            self.assert_endpoint_used(stubber.requests[0].url, endpoint)
+
+    def test_endpoint_discovery_default_optional_endpoint(self):
+        client, http_stubber = self.create_client(service_name="test-discovery-endpoint")
+        with http_stubber as stubber:
+            stubber.add_response(status=200, body=b'{}')
+            client.test_discovery_optional(Foo="bar")
+            self.assertEqual(len(stubber.requests), 1)
+
+    def test_endpoint_discovery_enabled_optional_endpoint(self):
+        discovered_endpoint = 'https://discovered.domain'
+        config = Config(endpoint_discovery_enabled=True)
+        client, http_stubber = self.create_client(
+            service_name="test-discovery-endpoint", config=config
+        )
+        with http_stubber as stubber:
+            self.add_describe_endpoints_response(stubber, discovered_endpoint)
+            client.test_discovery_optional(Foo="bar")
+            self.assert_endpoint_discovery_used(stubber, discovered_endpoint)
+
+    def test_endpoint_discovery_manual_auto_on_required_endpoint(self):
+        discovered_endpoint = 'https://discovered.domain'
+        config = Config(endpoint_discovery_enabled=" aUto  \n")
```

Comment:
It seems weird that we would support this casing with whitespaces especially when we are providing it through the `Config` object. I think it may make sense to not try to do any `lower` or `strip` for the auto value. I do not feel too strongly about this though.

Reply:
So I could move this to the config file, but I think it's a pretty common case to get weird spaces or extra return character in config files. I just want to assert we're doing case normalization. If we can't do an enum for this, we should be doing out best to avoid silly failures.

Reply:
Ok I'm fine with keeping it as is then.